### PR TITLE
GH-36795: [C#] Implement support for dense and sparse unions

### LINKED ISSUE #36795: [C#] Implement support for dense and sparse unions
### Describe the enhancement requested

Union is not currently supported from C#. Add it.

### Component(s)

C#

### PR DESCRIPTION
**Pull request by CurtHagenlocher**
### What changes are included in this PR?

Support dense and sparse unions in the C# implementation.
Adds Archery support for C# unions.

### Are these changes tested?

Yes

### Are there any user-facing changes?

Unions are now supported in the C# implementation.

**This PR includes breaking changes to public APIs.**

The public APIs for the UnionArray and UnionType were changed fairly substantially. As these were previously not implemented properly, the impact of the changes ought to be minimal.

The ChunkedArray and Column classes were changed to hold IArrowArrays instead of Arrays. To accomodate this, a constructor was added which may introduce ambiguity in calling code. This could be avoided by changing the overloaded constructor to instead be a factory method. This didn't seem worthwhile but could be reconsidered.

The metadata version was finally increased to V5.
 
* Closes: #36795

### CODE REVIEW

File: csharp/src/Apache.Arrow/Arrays/DenseUnionArray.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            _fields = children.ToArray();
+            ValidateMode(UnionMode.Dense, Type.Mode);
+        }
+
+        public DenseUnionArray(ArrayData data) 
+            : base(data)
+        {
+            ValidateMode(UnionMode.Dense, Type.Mode);
+            data.EnsureBufferCount(2); // TODO:
```
Comment (lidavidm) [inline csharp/src/Apache.Arrow/Arrays/DenseUnionArray.cs:49]: maybe link this TODO to a new issue? (and what is the TODO about, given SparseUnionArray lacks one?)

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Hmm, it seems C# producing unions fails in the integration test, though

File: csharp/src/Apache.Arrow/Arrays/DenseUnionArray.cs (hunk 1 above)
Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Arrays/DenseUnionArray.cs:49]: Heck if I can remember...

**Comment by CurtHagenlocher** [comment]
It looks like some changes made to Archery in the last week have fixed the errors we were getting. I'm ... not entirely comfortable with that, I guess, but I spent a lot of time trying to understand if/what the C# implementation was doing differently without any success.

**Comment by lidavidm** [comment]
Likely https://github.com/apache/arrow/commit/b957847a616a82e89af988c3c922af79379f6160? Which does now explicitly generate union fields with/without nullability to check what happens.

**Comment by lidavidm** [comment]
The Java error makes it sound like C# is still writing with metadata version 4: https://github.com/apache/arrow/blob/fa4310635c784f03fe825ecc818efa3eca361ec0/format/Schema.fbs#L38-L48

That may also explain things, if it's triggering some compatibility path, e.g.:
https://github.com/apache/arrow/blob/fa4310635c784f03fe825ecc818efa3eca361ec0/cpp/src/arrow/ipc/reader.cc#L391-L413

**Comment by CurtHagenlocher** [comment]
Thanks @lidavidm for your help; couldn't have made it work without you!

**Review by lidavidm — APPROVED** [review]

**Pull request merged**